epuck_ros2_cpp/src/driver.cpp

```diff
@@ -0,0 +1,453 @@
+
+    // Update variables
+    mPrevPositionX = positionX;
+    mPrevPositionY = positionY;
+    mPrevAngle = angle;
+    mPrevLeftWheelRaw = leftWheelRaw;
+    mPrevRightWheelRaw = rightWheelRaw;
+
+    // Pack & publish odometry
+    nav_msgs::msg::Odometry msg;
+    msg.header.stamp = stamp;
+    msg.header.frame_id = "odom";
+    msg.child_frame_id = "base_link";
+    msg.twist.twist.linear.x = v;
+    msg.twist.twist.linear.z = omega;
+    msg.pose.pose.position.x = positionX;
+    msg.pose.pose.position.y = positionY;
+    msg.pose.pose.orientation = euler2quaternion(0, 0, angle);
+    mOdometryPublisher->publish(msg);
+
+    // Pack & publish transforms
+    geometry_msgs::msg::TransformStamped tf;
+    tf.header.stamp = stamp;
+    tf.header.frame_id = "odom";
+    tf.child_frame_id = "base_link";
+    tf.transform.translation.x = positionX;
+    tf.transform.translation.y = positionY;
+    tf.transform.translation.z = 0.0;
+    tf.transform.rotation = euler2quaternion(0, 0, angle);
+    mDynamicBroadcaster->sendTransform(tf);
+  }
+
+  void updateCallback() {
+    int success;
+    int retryCount;
+    rclcpp::Time stamp;
+
+    // I2C: Set address
+    retryCount = 1;
+    success = 0;
+    while (!success && retryCount > 0) {
+      success = mI2cMain->setAddress(0x1F);
+      retryCount--;
+    }
+
+    // I2C: Write/Read
+    retryCount = 3;
+    success = 0;
+    while (!success && retryCount > 0) {
+      // Write
+      mMsgActuators[MSG_ACTUATORS_SIZE - 1] = 0;
+      for (int i = 0; i < MSG_ACTUATORS_SIZE - 1; i++)
+        mMsgActuators[MSG_ACTUATORS_SIZE - 1] ^= mMsgActuators[i];
+      success = (mI2cMain->writeData(mMsgActuators, MSG_ACTUATORS_SIZE) == MSG_ACTUATORS_SIZE);
+
+      // Read
+      success &= (mI2cMain->readData(mMsgSensors, MSG_SENSORS_SIZE) == MSG_SENSORS_SIZE);
+      char checksum = 0;
+      for (int i = 0; i < MSG_SENSORS_SIZE - 1; i++)
+        checksum ^= mMsgSensors[i];
+      success &= (checksum == mMsgSensors[MSG_SENSORS_SIZE - 1]);
+
+      retryCount--;
+    }
+
+    stamp = now();
+
+    if (success) {
+      publishDistanceData(stamp);
+      publishOdometryData(stamp);
+      mI2cMainErrCnt = 0;
+    } else {
+      mI2cMainErrCnt++;
+    }
+  }
+
+  OnSetParametersCallbackHandle::SharedPtr mCallbackHandler;
+
+  rclcpp::TimerBase::SharedPtr mTimer;
+  rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr mLaserPublisher;
+  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr mOdometryPublisher;
+  rclcpp::Publisher<sensor_msgs::msg::Range>::SharedPtr mRangePublisher[9];
+  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr mSubscription;
+
+  std::unique_ptr<tf2_ros::StaticTransformBroadcaster> mLaserBroadcaster;
+  std::unique_ptr<tf2_ros::TransformBroadcaster> mDynamicBroadcaster;
+  std::unique_ptr<tf2_ros::StaticTransformBroadcaster> mInfraredBroadcasters[9];
+
+  std::unique_ptr<I2CWrapper> mI2cMain;
+
+  int mFh;
```

Comment:
I swear it is the last one :wink: 
But we try to avoid abbreviation (https://github.com/cyberbotics/webots/wiki/Coding-Style#cs44-no-abbreviations), instead you can use `mFile` or `mFileHandle`